File: deepspace/EnemyStarShip.java
/*
 * @author: Jesús García León
 * @file: EnemyStarShip
 */
package deepspace;

/**
 * @brief Naves enemigas
 */
public class EnemyStarShip implements SpaceFighter{
    private float ammoPower;
    private String name;
    private float shieldPower;
    private Damage damage;
    private Loot loot;
    
    EnemyStarShip(String n, float a, float s, Loot l, Damage d){
        ammoPower = a;
        name = n;
        shieldPower = s;
        loot = l;
        damage = d;
    }
    
    EnemyStarShip(EnemyStarShip e){
        ammoPower = e.getAmmoPower();
        name = e.getName();
        shieldPower = e.shieldPower;
        damage = e.getDamage();
        loot = e.getLoot();
    }
    
    public float getAmmoPower(){
        return ammoPower;
    }
    
    public float getShieldPower(){
        return shieldPower;
    }
    
    public String getName(){
        return name;
    }
    
    public Loot getLoot(){
        return loot;
    }
    
    public Damage getDamage(){
        return damage;
    }
    
    EnemyToUI getUIversion(){
        return new EnemyToUI(this);
    }
    
    @Override
    public float fire(){
        return ammoPower;
    }
    
    @Override
    public float protection(){
        return shieldPower;
    }
    
    @Override
    public ShotResult receiveShot(float shot){
        if(shot > shieldPower)
            return ShotResult.DONOTRESIST;
        else
            return ShotResult.RESIST;
    }
    
    @Override
    public String toString() {
        String aux = "Nombre del enemigo: "+name;
        aux += "\nPoder de disparo: "+ammoPower;
        aux += "\nPoder de escudo: "+shieldPower;
        aux += "\nDaño de la nave:\n"+damage.toString();
        aux += "\nBotín de la nave:\n"+loot.toString();
        return aux;
    }
}

File: DeepspaceRuby/lib/EnemyStarShip.rb
#encoding: UTF-8

require_relative 'EnemyToUI'
require_relative 'ShotResult'

module Deepspace
    class EnemyStarShip
        @ammoPower
        @name
        @shieldPower
        @damage
        @loot

        def initialize(n,a,s,l,d)
            @ammoPower = a
            @name = n
            @shieldPower = s
            @damage = d
            @loot = l
        end

        def self.newCopy(e)
            return EnemyStarShip.new(e.name, e.ammoPower, e.shieldPower, e.loot, e.damage)
        end

        def ammoPower
            @ammoPower
        end

        def name
            @name
        end

        def shieldPower
            @shieldPower
        end

        def damage
            @damage
        end

        def loot
            @loot
        end

        def getUIversion
            return EnemyToUI.new(self)
        end

        def fire
            @ammoPower
        end

        def protection
            @shieldPower
        end

        def receiveShot(shot)
            if(shot > shieldPower)
                return ShotResult::DONOTRESIST
            else
                return ShotResult::RESIST
            end
        end

        def to_s
            string = "Name: #{@name}\n"
            string += "AmmoPower: #{@ammoPower}\n"
            string += "ShieldPower: #{@shieldPower}\n"
            if loot == nil
                string += "Loot: none\n"
            else
                string += "Loot: #{@loot.to_s}\n"
            end
            if damage == nil
                string += "Damage: none\n"
            else
                string += "Damage: #{@damage.to_s}\n"
            end 
        end
    end
end